nmostesting/mocks/Registry.py

```diff
@@ -271,3 +417,198 @@ def heartbeat(version, node_id):
         return jsonify({"health": int(time.time())})
     else:
         abort(404)
+
+
+@REGISTRY_API.route('/x-nmos/query', methods=["GET"], strict_slashes=False)
+def query_root():
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    base_data = ['v1.0/', 'v1.1/', 'v1.2/', 'v1.3/']
+
+    return Response(json.dumps(base_data), mimetype='application/json')
+
+
+@REGISTRY_API.route('/x-nmos/query/<version>', methods=["GET"], strict_slashes=False)
+def query(version):
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    base_data = ['devices/', 'flows/', 'nodes/', 'receivers/', 'senders/', 'sources/', 'subscriptions/']
+
+    return Response(json.dumps(base_data), mimetype='application/json')
+
+
+def compare_resources(resource1, resource2):
+    try:
+        return NMOSUtils.compare_resource_version(resource1['version'], resource2['version'])
+    except Exception as e:
+        print(e)
+        return 0
+
+
+@REGISTRY_API.route('/x-nmos/query/<version>/<resource>', methods=["GET"], strict_slashes=False)
+def query_resource(version, resource):
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    MIN_SINCE = "0:0"
+    MAX_UNTIL = NMOSUtils.get_TAI_time()
+
+    base_data = []
+
+    if request.args.get('paging.since') or request.args.get('paging.until'):
+        registry.pagination_used = True
+
+    since = request.args.get('paging.since') if request.args.get('paging.since') else MIN_SINCE
+    until = request.args.get('paging.until') if request.args.get('paging.until') else MAX_UNTIL
+    new_until = until
+
+    try:
+        valid_resource_types = ['device', 'flow', 'node', 'receiver', 'sender', 'source', 'subscription']
+
+        resource_type = resource.rstrip("s")
+
+        if resource_type not in valid_resource_types:
+            error_message = {"code": 404,
+                             "error": "Invalid resource",
+                             "debug": resource_type}
+            return Response(json.dumps(error_message), status=404, mimetype='application/json')
+
+        registry.query_api_called = True
+
+        # Check to see if resource is being requested as a query
+        if request.args.get('id'):
+            resource_id = request.args.get('id')
+            base_data.append(registry.get_resources()[resource_type][resource_id])
```

Comment:
Fair enough to include special support for requests like `/nodes?id={nodeId}` (nearly equivalent to `/nodes/{nodeId}`) but I think you ought to be careful to return `501 Not Implemented` if there are any _other_ query parameters (at least Basic Query Syntax ones and `query.rql`). Obviously the mock _does_ support `paging.since`/`.until` (and ought to support `paging.limit`), so maybe you need an allow-list...

Similarly in the `else` branch, other query parameters ought to be rejected with a `501`.

Reply:
Further discussion needed on permissiveness/restrictiveness of accepted query parameters